content/posts/2019-08-03---Going-from-ember-to-vue.md

```diff
@@ -0,0 +1,306 @@
+`methods` are the equivalent of `actions` in Ember -- if there is a function that is going to be called from the template, it is in the `methods` hash.
+Functions can live outside of that hash (obviously) but this tells you what you expect to interact with your markup.
+
+If you look at the `someMethod` function, you will see the last line calls `this.emit('something-that-bubbles');`.
+This is how we accomplish something similar to "Data Down, Actions Up" (DDAU).
+It's a bit different and shouldn't be used for every case, but basically, when we `emit` something, any parent component can listen to that event (in this case `something-that-bubbles`) and respond to it.
+This means that child component does not need to receive an action as a parameter.
+All it does is emit something, pass any properties along with it and then it stops caring.
+It's up to the parent component to listen to it and respond.
+
+One of the things I disliked with the actions helper in Ember was that a child component might not always want to bubble something up.
+Say you had a component like this in Ember: `<AuthorInfo @updateProfile={{action "updateProfile"}} />`.
+We pass down an action called `updateProfile` which allows the logged in user to update their profile.
+However, I want to use that same layout elsewhere when viewing a post, but it isn't editable.
+So, I simply don't pass down that action: `<AuthorInfo />`.
+If the user triggers that action, however, I will get an error like `this.updateProfile is not a function` (or something similar).
+So, you wrap stuff in if conditions to make sure that's not the case.
+
+With emitting, the child component won't hit this error, as all it does is emit something out and it's up to the parent to intercept it.
+Calling `this.emit('update-profile')` won't cause an error if parent component doesn't have a method for updating that profile.
+
+## Computed Properties
+
+Properties that needs to be available on the template, need to be declared in the hash returned by the `data` function.
+This ends up being really handy in documenting what shows up in your template.
+Computed properties are declared in the `computed` hash and behave very similarly.
+
+I love computed properties in Ember and was a bit miffed when there wasn't something similar in React.
+With Vue, you declare your computed properties as functions and Vue is smart enough to observe anything you declare within the function.
+So, we go from something like this Ember:
+
+```javascript
+fullName: computed('firstName', 'lastName', function() {
+  return `${this.firstName} ${this.lastName}`;
+})
+```
+
+To this in Vue:
+
+```javascript
+fullName() {
+  return `${this.firstName} ${this.lastName}`;
+}
+```
+
+Computed properties have the same issues in Vue as Ember, in that it won't observe nested keys more than one level deep and it has some issues with observing arrays.
```

Comment:
You can perfectly observe multi-level nested keys in Ember. What you cannot do is observe nested arrays, as that would be computationally too expensive.

Reply:
That is what I meant to specify `foo.@each.prop.name`.  Will update!